lnd_test.go

```diff
@@ -3801,69 +3778,110 @@ func testSpiderShortestPath(net *lntest.NetworkHarness, t *harnessTest) {
+		ctxt, _ = context.WithTimeout(ctxb, timeout)
+		updateFeeReq := &lnrpc.PolicyUpdateRequest{
+			BaseFeeMsat:   baseFee,
+			FeeRate:       float64(feeRate),
+			TimeLockDelta: timeLockDelta,
+			Scope: &lnrpc.PolicyUpdateRequest_ChanPoint{
+				ChanPoint: chanPoints[i],
+			},
+		}
+		if _, err := nodes[connections[i][0]].UpdateChannelPolicy(ctxt, updateFeeReq); err != nil {
+			t.Fatalf("unable to update chan policy: %v", err)
 		}
 	}
+		
+
+	// Wait for everyone to receive the channel update
+	for i := 0; i < numNodes; i++ {
+		waitForChannelUpdate(t, graphSubs[i], expectedMessages)
+	}
+
+	// Initialize seed random in order to generate invoices.
+	prand.Seed(time.Now().UnixNano())
+	var wg sync.WaitGroup
+	wg.Add(numPayIntents)
+	var succeededPays uint64
+	var failedPays uint64
 
-	errChan := make(chan error)
 	for i := 0; i < numPayIntents; i++ {
 		go func(i int) {
-			for j := 0; j < numInvoices; j++ {
-				if resp, err := payStreams[payIntents[i][0]].Recv(); err != nil {
-					errChan <- errors.Errorf("payment stream has"+
-						" been closed: %v", err)
-					return
-				} else if resp.PaymentError != "" {
-					errChan <- errors.Errorf("unable to send "+
-						"payment from %v to %v: %v",
-						nodeNames[payIntents[i][0]], nodeNames[payIntents[i][1]],
-						resp.PaymentError)
+			defer wg.Done()
+			timeToSleep := 1000000.0 / (baseRate * payRates[i])
+			timeout := time.After(time.Duration(testTime) * time.Second)
+			tick := time.Tick(time.Duration(timeToSleep) * time.Microsecond)
+			for {
+				select {
+				case <-timeout:
 					return
+				case <-tick:
+					// generate invoice and add to target node
+					preimage := make([]byte, 32)
+					_, err := rand.Read(preimage)
+					if err != nil {
+						t.Fatalf("unable to generate preimage: %v", err)
+					}
+					invoiceReq := &lnrpc.Invoice{
+						Memo:      "testing",
+						RPreimage: preimage,
+						Value:     int64(paymentAmt),
+					}
+					resp, err := nodes[payIntents[i][1]].AddInvoice(ctxb, invoiceReq)
+					if err != nil {
+						t.Fatalf("unable to add invoice: %v", err)
+					}
+
+					invoice := resp.PaymentRequest
+					sendReq := &lnrpc.SendRequest{
+						PaymentRequest: invoice,
+						SpiderAlgo: routing.ShortestPath,
+					}
+					payresp, err := nodes[payIntents[i][0]].SendPaymentSync(ctxb, sendReq)
+					if err != nil {
+						atomic.AddUint64(&failedPays, 1)
+					} else {
+						if payresp.PaymentError != "" {
+							atomic.AddUint64(&failedPays, 1)
+						} else {
+							atomic.AddUint64(&succeededPays, 1)
+						}
+					}
 				}
 			}
-			errChan <- nil
 		}(i)
 	}
-	
-	// Wait for each node receive their payments, and throw and error
-	// if something goes wrong.
-	maxTime := 60 * time.Second
-	for i := 0; i < 2; i++ {
-		select {
-		case err := <-errChan:
-			if err != nil {
-				t.Fatalf(err.Error())
-			}
-		case <-time.After(maxTime):
-			t.Fatalf("waiting for payments to finish too long "+
-				"(%v)", maxTime)
-		}
-	}
+	wg.Wait()
+
```

Comment:
Similarly, can you add a comment saying "Collecting statistics - success rate and volume or something like that?"